### ISSUE TITLE
Context menu gets too small depending on the cursor position

### ISSUE THREAD

**Issue by kapodamy** [microsoft/vscode#224515]
Type: <b>Bug</b>

Context menus style has changed with lastest vscode version and does not look ok...

VS Code version: Code 1.92.0 (b1c0a14de1414fcdaa400695b4db1c0799bc3124, 2024-07-31T23:26:45.634Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|Screen Size|1366x768|

</details>

version:
![new](https://github.com/user-attachments/assets/929ff4fb-c0df-46dd-a662-916b32a0ba0a)

previous:
![old](https://github.com/user-attachments/assets/0a45abf8-2d9d-44bb-ae7f-82c38b0ebbb3)

Steps to Reproduce:

* Open context menu

any setting to revert the new style?

**Comment by Scorg** [comment]
I have a similar problem, but also with fonts being bold:
![ttt](https://github.com/user-attachments/assets/0970b4ac-672c-4ebc-998f-89315ab1422a)

**Comment by lah7** [comment]
I was going to open a report too. I couldn't find any mention in the release notes, or any setting, so I think it _could_ be related to the Electron 30 upgrade.

Using Arch Linux / KDE with a 2160p monitor here (2.0 global scale). Context menus are way too padded and large here too.

| Before | After | 
| --- | --- |
| ![Editor - Before](https://github.com/user-attachments/assets/6af981cd-921e-4c70-82ec-7856d9ffdf0e) | ![Editor - After](https://github.com/user-attachments/assets/5d491a5d-d06f-4091-8fbb-0808d8a094bf)
| ![Explorer - Before](https://github.com/user-attachments/assets/46f9b442-a1e4-4724-bf9b-f32eca4ea196) | ![Explorer - After](https://github.com/user-attachments/assets/cadcf33f-7a03-489b-aacc-3aa29f1ea142) _(I admit the newer menu does position better horizontally to the left)_ 

Looks like the new menu goes either all above or all below the cursor position, hence the scrolling.

Would really love to see an option for the old (native?) menus back. If that's impossible, at least add customisation options to control the padding, font, colours to make the new menu more comfortable.

**Comment by medovina** [comment]
I'm using Ubuntu 24.04.  With this VS Code release the context menu font is uncomfortably large on my machine as well.  I'd also like to have an option to use the old menu style.

I think the title of this issue (currently "Context menu gets too small depending on the cursor position") should be more general, e.g. "Context menu has inappropriate font size".

**Comment by godoyLeonardo** [comment]
> I'm using Ubuntu 24.04. With this VS Code release the context menu font is uncomfortably large on my machine as well. I'd also like to have an option to use the old menu style.
> 
> I think the title of this issue (currently "Context menu gets too small depending on the cursor position") should be more general, e.g. "Context menu has inappropriate font size".

inappropriate font size and padding

**Comment by chapmanjacobd** [comment]
It seems like the new context menu loads a 100~200 ms slower too

**Comment by Vallek** [comment]
Please give us an option to revert it! I don't need or want tablet style menus in a desktop code editor)

**Comment by halfburnttoast** [comment]
Also having this problem with Linux Mint MATE 21.3. The menu items are way too large. The context menu is so large that the items can't fit on the screen. I've downgraded back to 1.91 for now.

![codeissue1](https://github.com/user-attachments/assets/402bcf10-7402-41e2-b48b-fb49805c3c0c)
![codeissue2](https://github.com/user-attachments/assets/3617fae5-e7f5-48ac-879e-27eb65b0bcd8)

**Comment by CGooden9** [comment]
Same here. Also, the menu is bright white for me, despite using the dark modern theme. Maybe that's because I'm using the mixed Linux Mint theme (most parts are white, some are black for contrast)? I feel that it should at least match up with the app theme, like the old one did (or have an option to revert back to the old version altogether).
![Screenshot from 2024-08-03 18-52-17](https://github.com/user-attachments/assets/94429bce-cffe-4960-bab7-fadf4b83c9d3)

**Comment by CGooden9** [comment]
I've figured out a temporary solution, at least on linux. Set window.titleBarStyle to "custom". This restores the context menus, though also changes other parts of the UI. I've attached some screenshots for comparison.
![Screenshot from 2024-08-03 19-46-05](https://github.com/user-attachments/assets/2fb847c5-10d8-4218-b69c-a9b9afac4704)
![Screenshot from 2024-08-03 19-49-25](https://github.com/user-attachments/assets/a1554683-9c8c-430d-867b-224adf9ef977)

**Comment by kapodamy** [comment]
upstream problem https://github.com/electron/electron/issues/42262

**Comment by GaryMatthews** [comment]
cheers @CGooden9  this is exactly the fix i was looking for to solve the context menu font size issue. now its a matter of excess padding but this puts most if not all menu items back into view instead of having to scroll to menu items on wasted screen real estate which is a frustrating experience.

**Comment by akalankasakalasooriya** [comment]
Same happened to me on ubuntu 22.04. KDE plasma
https://github.com/microsoft/vscode/issues/225033

**Comment by siferati** [comment]
Same problem for me since updating VS Code today.
It's basically the same shitty Chrome Refresh 2023 that people have been complaining about for the past few months.
I guess it has finally affected Electron and VS Code too...
-  https://github.com/brave/brave-browser/issues/37770
- https://www.reddit.com/r/chrome/comments/1dkluz5/new_ui_megathread_3/

**Comment by vladlabs** [comment]
> window.titleBarStyle to "custom

But there is a small issue here custom title bar is awful.

**Comment by metablaster** [comment]
Same issue, the menu and fonts are too big, it covers almost my entire 1920x1080 screen, sometimes so big there is scroll menu on the menu to scroll the menu.

It feels like dealing with touch screen or a phone rather than desktop PC.

![vscode](https://github.com/user-attachments/assets/cc316b87-2d08-4a1c-90c3-b526919a2337)

**Comment by tycode** [comment]
I'm affected by this too.

I just updated VS Code from 1.91.1 to 1.92.1.

Opening the menu for an extension in the extensions view used to easily fully fit on the screen, now I get a menu that unnecessarily scrolls, because of two things:

1. it's bigger, as mentioned here already
2. it now wants to insist it's either fully above or fully below my cursor, so when the thing I'm clicking on (to open the menu in the first place) is half-way down the screen it can only use half the screen height, whereas before it was perfectly happy to move up or down a bit in order to fit fully on screen.

**Can we please have an option to use native OS menus only?**
If that's not possible, then fully custom ones that either just have sensible metrics, or have metrics we can tweak ourselves, would work fine too. Anything that successfully avoids the questionable design choices of Chrome, please.

I run Debian Linux with MATE, I use the "Native title bar" option in VS Code, and I do not use Chrome. Firefox's menus are fine, and all other non-Electron programs are fine. VS Code's menus used to be fine since they were native, now they are significantly bigger than any other menu I will ever see on my PC.

PS. I would also appreciate such an option on Windows. I've been bugged by this there since much longer ago since Electron menus have been non-native on Windows for ages, but I mostly solved that problem by switching most of my computer use to Linux(!), so for me this has now been "un-solved" since this now affects Linux too.

![menu-comparison](https://github.com/user-attachments/assets/8657dec4-4428-4d70-8f09-9c17bce95210)

Edit: I have successfully rolled back to 1.91.1. For now I will stay on that version until this menu issue is resolved.

**Comment by RokeJulianLockhart** [comment]
This issue was rather difficult to locate, to the extent that I accidentally filed a duplicate: https://github.com/microsoft/vscode/issues/225055#issue-2453505666. I suggest that it be renamed to better reflect the underlying cause - that the context menus are unexpectedly large.

**Comment by tycode** [comment]
> > [#224515 (comment)](https://github.com/microsoft/vscode/issues/224515#issue-2443317768)
> 
> @kapodamy, this issue was rather difficult to locate, to the extent that I accidentally filed a duplicate: [#225055 (comment)](https://github.com/microsoft/vscode/issues/225055#issue-2453505666). I suggest that it be renamed to better reflect the underlying cause - that the context menus are unexpectedly large.

It's more the fact that the root cause (Chrome's design change ending up in Electron and now in VS Code) has two separate visible effects/symptoms: one is the menu is bigger, the other is that the size now depends on the cursor position when it didn't before. So both issues were described accurately, just from different points of view.

Hopefully my comparison screenshot will help to make this clear both to the VS Code team and to anyone else coming to report the same problem.

**Comment by firavoyage** [comment]
ubuntu 24
same
fuck god

**Comment by domdfcoding** [comment]
Starting seeing absolutely massive context menus on Ubuntu 20.04 after updating to 1.92.2 yesterday. Could read them from across the room

**Comment by Foresteam** [comment]
Having the same issue.
- VS Code Version: 1.92+ (1.92.2)
- Was fine on 1.91.1.24193-1
- OS Version: Arch Linux 6.6.47-1-lts, Wayland, Plasma 6.1.4, QT 6.7.2

The problem: menu fonts got bigger with an update, and now often don't fit in the screen. 
Before:
![2024-08-23_04-04](https://github.com/user-attachments/assets/2916f9f9-c3f8-4d8c-bf5d-6679a5ed9275)
Now:
![2024-08-23_04-02](https://github.com/user-attachments/assets/54b907d1-81a9-4c8c-aab8-5703119af928)

**Comment by RossBrunton** [comment]
Agreed that this is terrible. Even ignoring the appearance, the fact that it jumps around so much when scrolling makes it obnoxious.

I know that `window.titleBarStyle` fixes this (on Linux at least), but requires you to have the fancy titlebar thingie. Would it be possible to split out a seperate option for "use native context menus" or similar?

**Comment by BellCubeDev** [comment]
Reporting in from a fresh install of Insiders (1.93.0-insider 36e4ddb4dbdf5d41bbeedfce2db4834085de2b87) on a fresh install of Mint (Mint 22, kernel 6.8.0-41-generic). Changing `window.titleBarStyle` replaces the context menu as well which restores the behavior I'm used to but the massive menus on start are hard to work with.

**Comment by Lehdari** [comment]
Same problem here, Ubuntu 22.04 LTS. Super distracting to work with, menus extend almost the entire vertical screen space with some extensions. I don't like the custom style either.

**Comment by ra50** [comment]
Same here on Mint 21.3. Menus too large, text too bold. Setting window.titleBarStyle to custom fixes it. But that should not be necessary. The way menus look with custom title bar should be the same with native title bar.

**Comment by yomajo** [comment]
vscode folks, could you please stop messing with editor settings / appearance?! If something needs to be added - add an option for that niche community. I would rather spend time coding, that browsing issues, attempting to restore my productivity.

Ubuntu 22.04.4 LTS. Past several weeks context menu items dont fit my 1360 x 768 screen anymore. For file delete in file explorer after right click, you have to scroll.

**Comment by flying-sheep** [comment]
Can y’all please stop with the useless “me too” responses?

That’s why GitHub introduced the 👍 reaction, just click it.

Also @yomajo, please learn how to read issues. You’re missing the fact that there is an upstream issue. Find it, and stop blaming the VS Code devs for “messing” with something they didn’t touch.

**Comment by RokeJulianLockhart** [comment]
@flying-sheep, note that you can subscribe solely to issue closure and reopening, if that's desirable.

**Comment by kc9jud** [comment]
According to the upstream bug report https://github.com/electron/electron/issues/42262#issuecomment-2346578790, this is indeed part of Chrome Refresh 2023 and there is no clear path forward for Electron to resolve this.

**Comment by tycode** [comment]
I have proposed a possible way forward for anyone affected by this issue: #229053

This doesn't restore proper native OS menus to newer Electron (since that would probably be extremely challenging), but it _does_ permit you to use the VS Code custom menus while using the native OS titlebar. I personally find this an acceptable solution, on par with the experience of "VS Code 1.91.1 on Linux with native titlebar & menus", but I'd be interested to hear what the consensus is on whether this approach works for everyone else watching this discussion.

**Comment by ra50** [comment]
Without knowing the details of the Electron change, it's crazy that VS Code *can* show native looking menus with the *custom* window title bar, but *can't* show native looking menus and instead shows these extra large white menus with the *native* window title bar.  It should be possible to patch out whatever madness is forcing this from Electron.

**Comment by tycode** [comment]
@ra50 I agree that VS Code's "custom" menus look more like native menus than Electron menus do. However - and I say this as someone who's now actually inspected the code behind it - VS Code's "custom" menus are not actually native at all. They are all done in CSS & JS, so they would have either been designed by the VS Code team or a contributor, or by someone else who made some menu library that the VS Code team then decided to use. If they look native, that's just because whoever designed them did a good job of... making them look native.

"Patching" VS Code to show "native looking" (custom) menus while using the native title bar is _exactly_ what my issue (linked above) proposes, and _exactly_ what my pull request at #229385 implements (my apologies that I didn't leave a comment here earlier with a link to this). All that is needed now is for this (or something else that does the same job) to be merged and included in an official release.

**Comment by makirill** [comment]
Chrome recently implemented a menu change that has been met with mixed reactions. Developers reduced the font weight, making the menu appear less bold and visually smaller. Is it possible to replicate this change in Visual Studio Code as a temporary workaround?

**Comment by kolbma** [comment]
Where does this font come from in Electron?  
I mean when you have all different kind of problems with the font, some too small, the other (like me) much too big and other bold, then there should be some way to influence it.  
So how does Electron decide which font in which size and style to use?

**Comment by jamesrobb** [comment]
Running into this with Manjaro with Plasma 6.1.5.

**Comment by elamperti** [comment]
Same issue here — Manjaro, i3wm, VSCode v1.96

Enabling the custom window style solves the context menu but fucks up the title bar, since it tries to use the vscode title bar implementation (which takes a gigantic chunk of precious vertical space).

**Comment by lah7** [comment]
💡 **Tip for Linux KDE/KWin users:** You can force the native title bar with a custom title bar style (fixing menus) by adding a window rule like so:

* System Settings → Window Management → Window Rules → New
  * Window class (application) : `code Code`
  * Add:  "No titlebar and frame" : "Apply Initially" = "No"
      ![Image](https://github.com/user-attachments/assets/13ad569f-c699-47ff-98b2-0b674dcba2c1)


Optionally, use the [Custom CSS JS Loader extension](https://marketplace.visualstudio.com/itemdetails?itemName=be5invis.vscode-custom-css) to hide the application icon in the menu bar with CSS:
```css
.window-appicon {
    display: none;
}
```

End result:
![Image](https://github.com/user-attachments/assets/520aeee6-4084-4361-bd1d-a7ebccdde989)

Although, this may not be useful if you want the title bar and tab bar **only**. This is like having a menu bar & _poor man's toolbar_ (#41309) stitched together above the tabs. There's also double window controls until #229385 is merged (thanks @tycode for working on that!). For me, the native title bar is important as it's my middle click shortcut to lower windows.

I actually prefer VS Code's custom menus over the "native" ones [the earlier versions had](https://github.com/microsoft/vscode/issues/224515#issuecomment-2265127532) - that's probably because VS Code's menus have an empty icon column, even though VS Code doesn't populate them (#208482), this menu layout/spacing is more in line with the rest of the system.

I'm not sure if the other window compositing managers (marco, metacity, openbox, i3) have this ability if you wanted to use this workaround.

**Comment by glitterkitty** [comment]
> I've figured out a temporary solution, at least on linux. Set window.titleBarStyle to "custom". This restores the context menus, though also changes other parts of the UI. 

https://github.com/microsoft/vscode/issues/224515#issuecomment-2267264540

thx @CGooden9 
worked for me, too

**Comment by RokeJulianLockhart** [comment]
> https://github.com/microsoft/vscode/issues/224515#issuecomment-2545724749

@lah7, are you using XOrg 11? I believe it'll only work for that and XWayland, per [`bugs.kde.org/show_bug.cgi?id=452240`](https://bugs.kde.org/show_bug.cgi?id=452240#c10).

**Comment by lah7** [comment]
@RokeJulianLockhart Yes, I'm using X11. Thanks for the info. I've updated my previous comment.

A plus for the custom menus is that they can be recoloured to blend in with the rest of the system:

![](https://github.com/user-attachments/assets/372e7ff2-0d20-4e7b-a95c-72cd1e046499) 

```json
{
    "workbench.colorCustomizations": {
        // Colours for KDE Breeze
        "menu.background": "#25282c",
        "menu.foreground": "#eaebef",
        "menu.selectionBorder": "#47a8d9",
        "menu.selectionBackground": "#2f5469",
        "menu.selectionForeground": "#eaebef",
        "menu.separatorBackground": "#55585b",
    }
}
```

Electron (upstream) made their position clear that they won't diverge from Chromium to bring back the old menus. Chromium (their upstream) will need to bring them back.... _but I have some doubts._

* https://github.com/electron/electron/issues/42262#issuecomment-2544981851

I believe these PRs can be a way forward for VS Code - avoid Chromium's so-called "native" menus and have VS Code split their configuration (as @tycode proposed) so native window title bars are not tied under ~native~ _ugly Chromium_ menus. It seems that Electron for macOS is only one that truly uses the OS' "native" menu.

* https://github.com/microsoft/vscode/issues/229053
* https://github.com/microsoft/vscode/pull/229385
* https://github.com/electron/electron/pull/45033

**Comment by oc109** [comment]
With the insider update (https://github.com/microsoft/vscode/pull/245423) users can set "Window: Menu Style" separately apart from "Window: Title Bar Style".

Here is the ideal result, native title bar and custom menu:

![Image](https://github.com/user-attachments/assets/87087b53-df52-4ad4-8684-f3768b3c15f8)

**Comment by flying-sheep** [comment]
Both menu styles are custom on Linux. Ideally, Electron would allow using an actually native menu on Linux.

But for the purpose of this issue (wrongly sized menu) I guess this fixes it!

**Comment by beedell-roke** [comment]
> Both menu styles are custom on Linux. Ideally, Electron would allow using an actually native menu on Linux.

@flying-sheep, "native" doesn't mean much here. Do you mean that VSC should implement a Qt 6 and GTK 4 context menu? Chromium has integrations with both toolkits (although its Qt 5 or 6 integration is dreadful, and the GTK 3 one is substandard too), but I believe that that's specific to the browser's GUI, rather than something provided with the engine itself. I don't see that occurring without a separate issue filed for it, and I don't think it would be in scope for upstream Electron.

**Comment by ra50** [comment]
> With the insider update ([#245423](https://github.com/microsoft/vscode/pull/245423)) users can set "Window: Menu Style" separately apart from "Window: Title Bar Style".
> 
> Here is the ideal result, native title bar and custom menu:
> 
> ![Image](https://github.com/user-attachments/assets/87087b53-df52-4ad4-8684-f3768b3c15f8)

There's a bug with this update.  With Title Bar Style set to native and Menu Style set to custom (the setting you need to get normal sized menus), the Menu Bar Visibility setting of "classic" no longer works - no matter what you pick, the menu doesn't show, even with Alt.

**Comment by tycode** [comment]
> There's a bug with this update. With Title Bar Style set to native and Menu Style set to custom (the setting you need to get normal sized menus), the Menu Bar Visibility setting of "classic" no longer works - no matter what you pick, the menu doesn't show, even with Alt.

I tested this last night (see https://github.com/microsoft/vscode/pull/251113#issuecomment-2967884289) and it looks fine to me:

![Image](https://github.com/user-attachments/assets/08665801-106c-4d0f-a73c-1a86e4653de9)

with the following version: (which I guess is now one day out of date)

> Version: 1.102.0-insider
> Commit: 8ece07a028f7e5f0e9f41031828f60a2a36d7411
> Date: 2025-06-12T05:05:34.387Z
> Electron: 35.5.1
> ElectronBuildId: 11727614
> Chromium: 134.0.6998.205
> Node.js: 22.15.1
> V8: 13.4.114.21-electron.0
> OS: Linux x64 6.1.0-37-amd64

and settings.json:

```json
{
    "window.titleBarStyle": "native",
    "window.menuStyle": "custom",
    "workbench.navigationControl.enabled": false,
    "window.commandCenter": false,
    "chat.commandCenter.enabled": false,
    "workbench.layoutControl.enabled": false
}
```

(Menu Bar Visibility is at its default of "classic")

If there's a new bug (maybe a newly introduced regression?) then probably better to open a new issue for it, with full repro steps.

**Comment by ra50** [comment]
> > There's a bug with this update. With Title Bar Style set to native and Menu Style set to custom (the setting you need to get normal sized menus), the Menu Bar Visibility setting of "classic" no longer works - no matter what you pick, the menu doesn't show, even with Alt.
> 
> I tested this last night (see [#251113 (comment)](https://github.com/microsoft/vscode/pull/251113#issuecomment-2967884289)) and it looks fine to me:
> 
> ![Image](https://github.com/user-attachments/assets/08665801-106c-4d0f-a73c-1a86e4653de9)
> 
> with the following version: (which I guess is now one day out of date)
> 
> > Version: 1.102.0-insider
> > Commit: [8ece07a](https://github.com/microsoft/vscode/commit/8ece07a028f7e5f0e9f41031828f60a2a36d7411)
> > Date: 2025-06-12T05:05:34.387Z
> > Electron: 35.5.1
> > ElectronBuildId: 11727614
> > Chromium: 134.0.6998.205
> > Node.js: 22.15.1
> > V8: 13.4.114.21-electron.0
> > OS: Linux x64 6.1.0-37-amd64
> 
> and settings.json:
> 
> {
>     "window.titleBarStyle": "native",
>     "window.menuStyle": "custom",
>     "workbench.navigationControl.enabled": false,
>     "window.commandCenter": false,
>     "chat.commandCenter.enabled": false,
>     "workbench.layoutControl.enabled": false
> }
> 
> (Menu Bar Visibility is at its default of "classic")
> 
> If there's a new bug (maybe a newly introduced regression?) then probably better to open a new issue for it, with full repro steps.

Thanks for this, toggling `window.commandCenter` on and back off made the menu appear.  Must've just been a side effect of the exact order of enabling things.

**Comment by oc109** [comment]
> There's a bug with this update. With Title Bar Style set to native and Menu Style set to custom (the setting you need to get normal sized menus), the Menu Bar Visibility setting of "classic" no longer works - no matter what you pick, the menu doesn't show, even with Alt.

As you can see I use the compact menu bar to avoid it.

**Comment by tycode** [comment]
> Thanks for this, toggling `window.commandCenter` on and back off made the menu appear. Must've just been a side effect of the exact order of enabling things.

Having now updated 1 Linux PC and 3 Windows PCs to VS Code Stable 1.102 myself, I had this issue on one of the Windows PCs, and the same fix worked: toggling `window.commandCenter` on and back off. The other three PCs had no problem.

@oc109 does this fix work for you?

**Comment by beedell-roke** [comment]
> https://github.com/microsoft/vscode/issues/224515#issuecomment-3075776122

@tycode, since multiple people have experienced this, I suggest someone affected file an issue about it.
